internal/app_config.go

```diff
@@ -0,0 +1,127 @@
+package internal
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/pelletier/go-toml"
+)
+
+var DEF_CONFIG_ORDER = []string{
+	"/usr/lib/sota/conf.d",
+	"/var/sota/sota.toml",
+	"/etc/sota/conf.d/",
+}
+
+type cfgFile struct {
+	name string
+	path string
+
+	tree *toml.Tree
+}
+
+type AppConfig struct {
+	cfgs []*cfgFile
+}
+
+// NewAppConfig parses config files as per:
+// https://docs.ota.here.com/ota-client/latest/aktualizr-config-options.html#_how_toml_files_are_processed
+func NewAppConfig(configPaths []string) (*AppConfig, error) {
+	configsMap := make(map[string]string)
+
+	// Build up the files
+	for _, path := range configPaths {
+		if st, err := os.Stat(path); err == nil {
+			if st.IsDir() {
+				if entries, err := os.ReadDir(path); err == nil {
+					for _, entry := range entries {
+						if strings.HasSuffix(entry.Name(), ".toml") && entry.Type().IsRegular() {
+							configsMap[entry.Name()] = filepath.Join(path, entry.Name())
+						}
+					}
+				}
+			} else {
+				configsMap[st.Name()] = path
+			}
+		}
+	}
+
+	// Sort them alphabetically
+	keys := make([]*cfgFile, 0, len(configsMap))
+	for k, v := range configsMap {
+		keys = append(keys, &cfgFile{name: k, path: v})
+	}
+	sort.Slice(keys[:], func(i, j int) bool {
+		return keys[i].name < keys[j].name
+	})
+
+	var err error
+	for _, cfg := range keys {
+		cfg.tree, err = toml.LoadFile(cfg.path)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	cfg := AppConfig{cfgs: keys}
+	return &cfg, nil
+}
+
+func (c AppConfig) Get(key string) string {
+	for i := len(c.cfgs) - 1; i >= 0; i-- {
+		val := c.cfgs[i].tree.GetDefault(key, "").(string)
+		if len(val) > 0 {
+			return val
+		}
+	}
+	return ""
+}
+
+func (c AppConfig) GetDefault(key string, defval string) string {
+	val := c.Get(key)
+	if len(val) == 0 {
+		val = defval
+	}
+	return val
+}
+
+func (c AppConfig) Write(keyVals map[string]string) error {
```

Comment:
This write won't work if the most significant file is located in the `/usr` (readonly fs).

Reply:
Maybe appconfig should read/get an overall/merged configuration while allow writing only to a specific file(s). If it happens that the value to be set is defined in some more significant and read only config file then just return an error.

Reply:
There is only one place we use this API: updating pkey/client-cert when doing a rotation. I thought about trying to do a specific file, but the point of this change was kind of because of the fact we don't know which file they might use. This can only work if the file is /var/sota/sota.toml or lives under /etc.

Maybe there's a better approach than creating a generic "write" method?

Reply:
Foremost, @doanac I'm fine with your justification.

Unrelated to this specific change, but related in general.

I guess some of our users would want to put their `sota.toml` of a read-only partition.
For them, we still need to provide an ability to rotate the client certificate.

One such way is to store the sota config client cert paths in a separate file.
So, maybe, we need to consider making that an option in e.g. `lmp-device-register`?

Reply:
My understanding is that even the current `Write` implementation will return an error if a client tries to set a value stored in the read-only partition anyway. So, it kind of already works in the way I described above.

Reply:
I'm still on the fence with this Write function. Part of me wants to only allow writing to `<prefix>/sota.toml` and error out if a more significant file would override it. But you could also try and keep the write and make it a little smarter to check for read-only errors and finally have some sort of "/etc/z99-fioconfig-hack.toml" as a fail-safe.

I'll think more on it, but I'm open to some alternatives if you have any

Reply:
> Foremost, @doanac I'm fine with your justification.
> 
> Unrelated to this specific change, but related in general.
> 
> I guess some of our users would want to put their `sota.toml` of a read-only partition. For them, we still need to provide an ability to rotate the client certificate.
> 
> One such way is to store the sota config client cert paths in a separate file. So, maybe, we need to consider making that an option in e.g. `lmp-device-register`?

I am not sure it will be easy to implement it. aklite/libaktualizr makes too many assumptions about the certs and sota.toml location. We also have this https://github.com/foundriesio/fioconfig/blob/1b1929bff291fa440e7efe7a29dfed9625b60ade/contrib/aktualizr-toml-update#L24 which assumes that sota.toml is writable.

Reply:
If `import.tls_*` are set in a read-only file it kind of assumes that a customer wouldn't like to support cert rotation. Therefore, we need to make it impose the requirement the certs' attributes should be defined in a writable `*.toml` file.

Now, should  the `import.tls_*` and `tls.*_source` attributes be defined only in `sota.toml` or they can be defined in one or more `*.toml` in `/etc`?

IMO, having those attributes defined in more than one `*.toml` file can create lots of confusion and lead to issues.
So, IMO, I would restrict writing the cert related attributes only to one file, either `sota.toml` or maybe a new one in `/etc`, not sure the later option is practical though.

In general, these generic `*.toml` mechanism is too generic :). There are certain critical parameters that should be defined only in a one `*.toml` file and we shouldn't allow overriding them. For example, does it make sense to allow overriding URLs specified in `sota.toml`?

I would say that this generic mechanism is a great thing for overriding/customizing configuration at the build time (yocto recipes in meta-subs), let's call it a static configuration.

When it comes to dynamic (re-)configuration, overriding in generic way is too dangerous. It has to be restricted and specific, as we do with `z-50-fioctl.toml`.

Reply:
What complicates this is that we have a generic platform that people can do almost anything with. Just the "single sota.toml" constraint we'd like: A customer might introduce a `/usr/lib/sota/conf.d/sota.toml` file they intend for global common stuff *and* have `/var/sota/sota.toml`. Or mind blown: `/var/sota-customized-for-no-good-reason/sota.toml`. So just knowing which sota.toml we should write to isn't foolproof.

Reply:
And the parsing order doesn't give us clues. i.e. the first entry by default is /usr/lib and not /var/sota.

Reply:
I think we need some kind of top level config which is defined solely in yocto (factory's lmp-manifest, meta-subs). This top-level config will define the things like this, e.g.

1. A list of dirs/files containing `*.toml` snippets for aklite config.
2. A path to the `*.toml` file that contains mTLS/certs related config.

Reply:
k. i'm revisiting this finally. I think I have a compromise. The interface to update TOML settings could have 2 new things added:

1. If the file we find to write values to isn't writeable, then we move on and try to find a more significant file that is writable.
2. Create some type of "Can Update Config" method that takes a set of keys. This will allow us to test up front before trying to do the entire key rotation.

Reply:
I think it should work.

As for the item 1. If more significant writable file is not found then the method/function will return an error?

Reply:
yep. that's my thinking